ExecuteFromOutsideV2

MEJIOMAH17 opened the pull request:
## Describe your changes

Added ability to execute transactions https://starknetjs.com/docs/next/guides/outsideExecution/


## Breaking changes

- [x] This issue contains breaking changes

<!-- List all breaking changes introduced by this issue -->
open interface Account has new method. (Library user have to implement it)

MEJIOMAH17 committed:
ExecuteFromOutsideV2

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/account/Account.kt:
```diff
@@ -346,6 +346,20 @@ interface Account {
         skipValidate: Boolean,
     ): Request<EstimateFeeResponseList>
 
+    /**
+     * @param caller: Authorized executor of the transaction(s):  Hex address or Felt.fromShortString("ANY_CALLER")
+     * @param executeAfter: Unix second timestamp of the beginning of the timeframe
+     * @param executeAfter: Unix second timestamp of the end of the timeframe
+     * @param calls: calls
+     */
```
1. [KDoc](https://kotlinlang.org/docs/kotlin-doc.html) doesn't use semicolon (`:`)
2. We're missing `nonce` param 😅 
```suggestion
    /**
     * @param caller Authorized executor of the transaction(s): Hex address or Felt.fromShortString("ANY_CALLER")
     * @param executeAfter Unix second timestamp of the beginning of the timeframe
     * @param executeAfter Unix second timestamp of the end of the timeframe
     * @param calls List of calls to be executed by outside caller
     * @param nonce Nonce for the transaction
     */
```

franciszekjob replied:
Also, I think the name of method could be simplified, e.g. `signOutsideExecutionCall`
Wdyt @MEJIOMAH17 ?

MEJIOMAH17 replied:
I have committed fixes based on your comments. [Remove : from KDoc](https://github.com/software-mansion/starknet-jvm/pull/552/commits/178875a9330b495a461998e544884d4a2301c498)

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/account/Account.kt:
I think we should also add following signatures (we will use methods overloading too):
- without `nonce` - nonce value for outside execution is not trivial, so we definitely need to allow users to populate it automatically
```kotlin
fun signOutsideExecutionCall(
    caller: Felt,
    executeAfter: Felt,
    executeBefore: Felt,
    calls: List<Call>,
): Call
```

- with single call
```kotlin
fun signOutsideExecutionCall(
    caller: Felt,
    executeAfter: Felt,
    executeBefore: Felt,
    calls: Call,
): Call
```

MEJIOMAH17 replied:
I have [added](https://github.com/software-mansion/starknet-jvm/pull/552/commits/b1d0dd5cf9281ecbaca98ea66cfbf9ba393d9a8d) overloads

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/account/StandardAccount.kt:
```diff
@@ -363,6 +364,48 @@ class StandardAccount @JvmOverloads constructor(
         }
     }
 
+    /**
+     * @param caller: Authorized executor of the transaction(s):  Hex address or Felt.fromShortString("ANY_CALLER")
+     * @param executeAfter: Unix second timestamp of the beginning of the timeframe
+     * @param executeAfter: Unix second timestamp of the end of the timeframe
+     * @param calls: calls
+     */
```
Same remark about KDoc as above.

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/account/StandardAccount.kt:
```diff
@@ -363,6 +364,48 @@ class StandardAccount @JvmOverloads constructor(
+        )
+        return Call(
+            contractAddress = address,
+            "execute_from_outside_v2",
```
Why don't we add support for both interface versions (v1, v2)?

MEJIOMAH17 replied:
We (extended) just do not need it. If you insist, I can add support for V1.

franciszekjob replied:
If that won't take much time, let's add it now. Otherwise, let's create an issue for that and link it.

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/data/types/OutsideCall.kt:
```diff
@@ -0,0 +1,37 @@
+
+@Serializable
+data class OutsideCall(
+    @SerialName("To")
+    val to: Felt,
+    @SerialName("Selector")
+    val selector: Felt,
+    @SerialName("Calldata")
+    val calldata: List<Felt>,
```
nit: readability: 
```suggestion
    @SerialName("To")
    val to: Felt,
    
    @SerialName("Selector")
    val selector: Felt,
    
    @SerialName("Calldata")
    val calldata: List<Felt>,
```

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/data/types/OutsideCall.kt:
```diff
@@ -0,0 +1,37 @@
+    val calldata: List<Felt>,
+) : ConvertibleToCalldata {
+    companion object {
+        val typeDescriptor: Descriptor = Descriptor(
```
nit:
```suggestion
        val typeDescriptor = Descriptor(
```

MEJIOMAH17 replied:
Fixed https://github.com/software-mansion/starknet-jvm/pull/552/commits/c148d56b30da212c5280a96d587f866b5cdc6ede

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/data/types/OutsideExecution.kt:
```diff
@@ -0,0 +1,77 @@
+
+@Serializable
+data class OutsideExecution(
+    @SerialName("Caller")
+    val caller: Felt,
+    @SerialName("Nonce")
+    val nonce: Felt,
+    @SerialName("Execute After")
+    val executeAfter: Felt,
+    @SerialName("Execute Before")
+    val executeBefore: Felt,
+    @SerialName("Calls")
+    val calls: List<OutsideCall>,
```
nit: readability:
```suggestion
    @SerialName("Caller")
    val caller: Felt,
    
    @SerialName("Nonce")
    val nonce: Felt,
    
    @SerialName("Execute After")
    val executeAfter: Felt,
    
    @SerialName("Execute Before")
    val executeBefore: Felt,
    
    @SerialName("Calls")
    val calls: List<OutsideCall>,
```

MEJIOMAH17 replied:
Fixed https://github.com/software-mansion/starknet-jvm/pull/552/commits/c148d56b30da212c5280a96d587f866b5cdc6ede

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/data/types/OutsideExecution.kt:
```diff
@@ -0,0 +1,77 @@
+    fun toTypedData(chainId: StarknetChainId): TypedData {
+        val domain = domain(chainId)
+
+        val OutsideExecutionTypesV2 = listOf(
```
(same for other occurrences below):
```suggestion
        val outsideExecutionTypesV2 = listOf(
```

MEJIOMAH17 replied:
Fixed https://github.com/software-mansion/starknet-jvm/pull/552/commits/c148d56b30da212c5280a96d587f866b5cdc6ede

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
```diff
@@ -249,7 +262,8 @@ class StandardAccountTest {
             val feeEstimate = request.send().values.first()
             // docsEnd
             assertNotEquals(Felt.ZERO, feeEstimate.overallFee)
-            val calculatedFee = feeEstimate.l1GasPrice.value * feeEstimate.l1GasConsumed.value + feeEstimate.l1DataGasPrice.value * feeEstimate.l1DataGasConsumed.value + feeEstimate.l2GasPrice.value * feeEstimate.l2GasConsumed.value
+            val calculatedFee =
+                feeEstimate.l1GasPrice.value * feeEstimate.l1GasConsumed.value + feeEstimate.l1DataGasPrice.value * feeEstimate.l1DataGasConsumed.value + feeEstimate.l2GasPrice.value * feeEstimate.l2GasConsumed.value
```
Not sure if this change is needed 😅

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
```diff
@@ -345,7 +359,8 @@ class StandardAccountTest {
             // docsEnd
             assertEquals(TransactionVersion.V3_QUERY, declareTransactionPayload.version)
             // docsStart
-            val request = provider.getEstimateFee(payload = listOf(declareTransactionPayload), simulationFlags = emptySet())
+            val request =
+                provider.getEstimateFee(payload = listOf(declareTransactionPayload), simulationFlags = emptySet())
```
Same as above

MEJIOMAH17 replied:
Fixed https://github.com/software-mansion/starknet-jvm/pull/552/commits/c148d56b30da212c5280a96d587f866b5cdc6ede

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
```diff
@@ -1297,4 +1312,73 @@ class StandardAccountTest {
+            ).send()[0].value
+        }
+
+        private fun transferCall(amount: Int): Call {
```
(same for occurrences above):
```suggestion
        private fun createTransferCall(amount: Int): Call {
```

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
```diff
@@ -1297,4 +1312,73 @@ class StandardAccountTest {
+            val call2 = transferCall(200)
+            val balanceBefore = getBalance(account.address)
+            val now = Instant.now()
+            val outsideCall = account.createExecuteFromOutsideV2Call(
+                caller = secondAccount.address,
+                executeAfter = Felt(now.minusSeconds(10).epochSecond),
+                executeBefore = Felt(now.plusSeconds(10).epochSecond),
+                calls = listOf(call1, call2),
+                nonce = Felt(Random.nextLong().absoluteValue),
+            )
```
Once we add signature without `nonce`, there won't be need to pass it here.

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
```diff
@@ -1297,4 +1312,73 @@ class StandardAccountTest {
+
+    @Nested
+    inner class OutsideExecutionTest {
+        private val randomAddress = Felt.fromHex("0x655b2696f187daec1952ef59befbd875add757748d6c844fb273439cd0d82f9")
```
If this is a purely random (not actual one) address, then let's use a dummy value (e.g. `0x123`).

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
Test are looking good, but we need a couple more cases 😄 :
- test with impossible time bounds (execute after is greater than execute before)
- test for signing by our account
- test where `caller` is a random address (for argent accounts it will return `argent/invalid-caller` error, not sure how for OZ, could you check this please?)

MEJIOMAH17 replied:
I added the tests https://github.com/software-mansion/starknet-jvm/pull/552#:~:text=Invalid%20address%20for%20tests.

franciszekjob reviewed (changes requested):
Overall looks good 👍 

Key remarks:
1. ATM there is support for v2, I think v1 should be included as well.
2. We should allow the nonce to be automatically generated.
3. We should mention outside execution in the docs (you can surely add it in separate PR)

Let me know if you need any clarification!

franciszekjob commented:
I see CI failed - once you remove these unnecessary formatting changes it should pass 👌

MEJIOMAH17 committed:
Remove : from KDoc
Add nonce parameter to the KDoc
createExecuteFromOutsideV2Call -> signOutsideExecutionCall

MEJIOMAH17 committed:
Add overloads for signOutsideExecutionCall

MEJIOMAH17 committed:
Invalid address for tests.
Add test executes transfers from the same account
Add test does not execute transfers signed to the different account
Add test does not execute transfers after allowed time interval

MEJIOMAH17 committed:
Readability

MEJIOMAH17 committed:
Refactor to support outside_execution_v1

MEJIOMAH17 committed:
Add outsideExecution to doc

MEJIOMAH17 committed:
Formatting

MEJIOMAH17 committed:
generateGuides

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/account/Account.kt:
```diff
@@ -346,6 +354,86 @@ interface Account {
+            executeAfter = executeAfter,
+            executeBefore = executeBefore,
+            calls = listOf(call),
+            nonce = Felt.random(),
```
```suggestion
            nonce = nonce,
```

MEJIOMAH17 replied:
Thank you!

franciszekjob commented on lib/guide.md:
```diff
@@ -859,6 +859,22 @@ val typedData = TypedData.fromJsonString("...")
+// user account, as long as they have the relevant signatures
+val secondAddress = Felt(0x1235)
+val secondPrivateKey = Felt(0x2)
+val secondAccount: Account = StandardAccount(secondAddress, secondPrivateKey, provider, chainId)
```
```suggestion
val secondAccount = StandardAccount(secondAddress, secondPrivateKey, provider, chainId)
```

franciszekjob commented on lib/guide.md:
```diff
@@ -859,6 +859,22 @@ val typedData = TypedData.fromJsonString("...")
+val now = Instant.now()
+val outsideCall = account.signOutsideExecutionCallV2(
+    caller = secondAccount.address,
+    executeAfter = Felt(now.minusSeconds(10).epochSecond),
+    executeBefore = Felt(now.plusSeconds(10).epochSecond),
```
nit: It's not important but can we lengthen the time window of possible execution (e.g. 1 hour instead of 10 seconds) 😅 ?

MEJIOMAH17 committed:
Update lib/src/main/kotlin/com/swmansion/starknet/account/Account.kt

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

MEJIOMAH17 committed:
Update lib/guide.md

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

MEJIOMAH17 committed:
Update lib/src/main/kotlin/com/swmansion/starknet/account/Account.kt

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

MEJIOMAH17 committed:
Fix invalid nonce.

Fix formatting

MEJIOMAH17 committed:
Drop Felt.random.

MEJIOMAH17 commented on lib/src/main/kotlin/com/swmansion/starknet/data/types/Felt.kt:
done

MEJIOMAH17 committed:
generateGuides

franciszekjob commented on lib/src/main/kotlin/com/swmansion/starknet/account/StandardAccount.kt:
```diff
@@ -363,6 +364,90 @@ class StandardAccount @JvmOverloads constructor(
+        )
+    }
+
+    override fun signOutsideExecutionCallV2(
+        caller: Felt,
+        executeAfter: Felt,
+        executeBefore: Felt,
+        calls: List<Call>,
+    ): Call {
+        return signOutsideExecutionCallV2(
+            caller = caller,
+            executeAfter = executeAfter,
+            executeBefore = executeBefore,
+            calls = calls,
+            // generatePrivateKey == random felt
+            nonce = generatePrivateKey(),
+        )
+    }
```
Last thing about nonce generation:
1. let's put into a to separate method (have a look [here](https://github.com/software-mansion/starknet.py/blob/e72e4c7b67f6faf851475f2f3eeda42aa7776b59/starknet_py/net/account/account.py#L255) for reference, it's from starknet.py SDK)
2. We need to check the nonce by `is_valid_outside_execution_nonce` function (reference [here](https://github.com/software-mansion/starknet.py/blob/e72e4c7b67f6faf851475f2f3eeda42aa7776b59/starknet_py/net/account/account.py#L237))

MEJIOMAH17 replied:
I have added getOutsideExecutionNonce and isValidOutsideExecutionNonce methods.

MEJIOMAH17 committed:
Add Account.getOutsideExecutionNonce

franciszekjob commented on lib/src/test/kotlin/starknet/account/StandardAccountTest.kt:
```diff
@@ -1297,4 +1310,168 @@ class StandardAccountTest {
         }
     }
+
+    @Nested
+    inner class OutsideExecutionTest {
+        private val randomAddress = Felt.fromHex("0x123")
```
I don't think these tests should be wrapped in inner class - note that they belong to `StandardAccount`, so there's no need to wrap it in extra class.

MEJIOMAH17 commented on lib/src/main/kotlin/com/swmansion/starknet/account/StandardAccount.kt:
```diff
@@ -198,25 +199,7 @@ class StandardAccount @JvmOverloads constructor(
-                    return@handle result[0] > Felt.ZERO
-                }
-            }
-        }
+        return request.toBooleanRequest()
```
To reuse same logic of converting Request<FeltArray> -> Request<Boolean> between methods. There is no need for copypaste.

franciszekjob commented on lib/guide.md:
```diff
@@ -859,6 +859,23 @@ val typedData = TypedData.fromJsonString("...")
 val typedDataSignature = account.signTypedData(typedData)
 val isValidSignatureRequest = account.verifyTypedDataSignature(typedData, typedDataSignature)
 val isValidSignature = isValidSignatureRequest.send()
+
+// Outside Execution, also known as meta-transactions, allows a protocol to submit transactions on behalf of a 
+// user account, as long as they have the relevant signatures
+val secondAddress = Felt(0x1235)
+val secondPrivateKey = Felt(0x2)
+val secondAccount = StandardAccount(secondAddress, secondPrivateKey, provider, chainId)
+
+val now = Instant.now()
+val outsideCall = account.signOutsideExecutionCallV2(
+    caller = secondAccount.address,
+    executeAfter = Felt(now.minus(Duration.ofHours(1)).epochSecond), 
+    executeBefore = Felt(now.plus(Duration.ofHours(1)).epochSecond),
+    calls = listOf(call),
+    nonce = getOutsideExecutionNonce()
+)
+
+secondAccount.executeV3(outsideCall).send()
```
```suggestion

// Outside Execution, also known as meta-transactions, allows a protocol to submit transactions on behalf of a 
// user account, as long as they have the relevant signatures
val secondAddress = Felt(0x1235)
val secondPrivateKey = Felt(0x2)
val secondAccount = StandardAccount(secondAddress, secondPrivateKey, provider, chainId)

val now = Instant.now()

// create the call for outside execution
val outsideCall = account.signOutsideExecutionCallV2(
    caller = secondAccount.address,
    // transaction can be executed within the following time window
    executeAfter = Felt(now.minus(Duration.ofHours(1)).epochSecond), 
    executeBefore = Felt(now.plus(Duration.ofHours(1)).epochSecond),
    calls = listOf(call),
    nonce = getOutsideExecutionNonce()
)

// execute the transaction from another account
secondAccount.executeV3(outsideCall).send()
```

franciszekjob replied:
Also, could we also have an example for any caller?

franciszekjob reviewed (commented):
Last thing and we're good to go 👍

MEJIOMAH17 committed:
Update lib/guide.md

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

MEJIOMAH17 committed:
Add example of OutsideExecution by ANY_CALLER
